### DISCUSSION TITLE
Mediawiki - Support for multiple seperate wikis

### DISCUSSION THREAD

**Discussion by Ernstian** [MichaIng/DietPi#7236] [Q&A]
Mediawiki is already supported, but there are two problems that I noticed while using it:

1. The smaller problem: the path /wiki/ is hardcoded, and I'm not sure if I can change it without breaking stuff. This is a problem if I want to use ["Short URLS"](https://www.mediawiki.org/wiki/Manual:Short_URL) (in short: Currently, it's "https://example.org/wiki/index.php?title=Page_title", it's possible to make it more accessible it being "https://example.org/wiki/Page_title" (like Wikipedia has it), but that needs a different installation path than /wiki/)
2. The main thing: I'm using the wiki for my pen and paper campaign, which works incredibly well (if you run a long-term campaign, try a wiki!). I'm thinking about starting another campaign soon, for which I want to provide another wiki. Installing another wiki does not work with dietpi out of the box.

I'm not sure how feasable it is, but I'd like to choose a "wikiname" or something during installation. The database and the website path should reflect that name and should be fully seperate (so I could uninstall one wiki, but keep the other intact): "https://example.org/wiki1/" and "https://example.org/wiki2/", for example. Something like a [wiki family with "Separate settings files"](https://www.mediawiki.org/wiki/Manual:Wiki_family) sounds like it would perfectly fit my use case.

**Comment by MichaIng** [comment]
@Ernstian
Why does the `/wiki` path breaks the short URLs? Following the guide for e.g. Lighttpd:
```sh
url.rewrite-if-not-file = ( 
    "^/wiki/(mw-)?config/?" => "$0",
    "^/wiki/([^?]*)(?:\?(.*))?" => "/w/index.php?title=$1&$2",
    "^/wiki/([^?]*)" => "/w/index.php?title=$1",
    "^/wiki$" => "/w/index.php", # to avoid 404 when the user types /wiki instead of /wiki/
)
```
So there the base installation is in `/w`. With it being in `/wiki`, the following should do:
```sh
url.rewrite-if-not-file = ( 
    "^/wiki/(mw-)?config/?" => "$0",
    "^/wiki/([^?]*)(?:\?(.*))?" => "/wiki/index.php?title=$1&$2",
    "^/wiki/([^?]*)" => "/wiki/index.php?title=$1",
    "^/wiki$" => "/wiki/index.php", # to avoid 404 when the user types /wiki instead of /wiki/
)`
```
Just replacing `/w` with `/wiki`. This works because `url.rewrite-if-not-file` only applies for URLs which do not point to a file (like `/wiki/index.php`) yet. Actually this way of shorting URLs is the more common one, using the base installation path, just removing the `index.php` and in case turning arguments into path elements.

We could actually implement these shortings OOTB to your webserver drop-ins. I guess, on Apache, it does work OOTB already via `.htaccess`.

It would be indeed not so difficult to implement a custom path/wiki name, with an optional dialogue and `dietpi.txt` entry for automation. But a problem for multiple installations is that `dietpi-software` knows only installed and no installed. It would still work with a reinstall and choosing another name/path, but e.g. for updating, again the path would need to be changed back. It would work, but somehow feels hacky/not right. Could be made a little better with some auto-detection of existing MediaWiki instances, checking for some `/var/www/*/<key_file>` that is unique to MediaWiki, then, in case, showing a menu to select the instance which shall be reinstalled/updated.

Btw, does MediaWiki have an internal updater, or do files need to be replaced manually?

**Reply by Ernstian** [reply]
>Actually this way of shorting URLs is the more common one, using the base installation path, just removing the index.php and in case turning arguments into path elements.

I wasn't aware that that was possible. I assumed it'd break something if it's the same path. Thank you for the correction! I still have much to learn 😄 I'll try that later.

>Could be made a little better with some auto-detection of existing MediaWiki instances, checking for some` /var/www/*/<key_file>` that is unique to MediaWiki, then, in case, showing a menu to select the instance which shall be reinstalled/updated.

With the rewrite rules, couldn't multiple installations be stored in `/var/www/wiki/` , for example `/var/www/wiki/wiki1` and `/var/www/wiki/wiki2`, with a rewrite to redirect `/wiki1/page` to `/wiki/wiki1/page` and `/wiki2/page` to `/wiki/wiki2/page`, making it easy for the user to access pages, and making it easy for the system to find all instances?

>Btw, does MediaWiki have an internal updater, or do files need to be replaced manually?

As far as I can tell, every update needs to be manual. The only internal updater is for the database, but that script has to be manually started, too ( [update.php in /maintenance](https://www.mediawiki.org/wiki/Manual:Update.php) ) (which could be added to the Dietpi reinstallation script, if it's not there already)

**Reply by MichaIng** [reply]
Regarding `/var/www/wiki/wiki1...` I think having them correctly named in the document root is better, as then it does not solely rely on rewrite rules, but the instance can be accessed via `domain.org/name` in any case. That way also we do not need to perform any migration for existing installs, which are not distributed in sub directories, yet + create rewrite rules etc.

There is actually a web updater as well, but it is not recommended, as it is a single script, which can exceed the PHP execution limit: https://www.mediawiki.org/wiki/Manual:Upgrading#Web_updater

Okay, actually something we should add to your docs: reinstall first, then running the database update via CLI.

**Reply by Ernstian** [reply]
I just noticed why I thought it's a problem if the wiki is located in /wiki: There's a [warning on the Short URLs page](https://www.mediawiki.org/wiki/Manual:Short_URL#Moving_a_wiki_from_/wiki_to_/w) (which includes a guide to move the wiki)

>A common beginner's mistake is to install MediaWiki itself (the source code, not the short URL) in /wiki instead of /w. Once installed they would realize the mistake when trying to configure Short URLs (which would become difficult as the virtual path then conflicts with the real path). 

There also is a warning on the [lighttpd guide](https://www.mediawiki.org/wiki/Special:MyLanguage/Manual:Short_URL/wiki/Page_title_--_Lighttpd_rewrite--root_access): 

>Please note the installation path (/w/ above) and the virtual path to access articles (/wiki/ above) must be different, or the styles won't load at all (and other issues, with extensions for instance), as explained [in the main article](https://www.mediawiki.org/wiki/Manual:Short_URL#Setup_steps)

I guess I'll stay away from this for now - never change a running system :) The bigger problem for me way the multiple wiki system anyway. Thank you for looking into it!

**Reply by MichaIng** [reply]
This is only because of their particular configs which do the rewrite even if the file exists, and in case of Lighttpd I do not see the issue at all. But I will do some tests.

**Reply by Ernstian** [reply]
Thank you so much for your work.

I took a look at the installation script for mediawiki in dietpi-software. Am I understanding right that, for creating a new mediawiki installation after having one installed by dietpi, the only steps I need to take are:

1. Create a new database table:
`/boot/dietpi/func/create_mysql_db newwikiname mediawiki masterpassword`
2. Curl mediawiki, export it to `/var/www/newwikiname` and chown to www-data

And then go through the normal setup wizard by mediawiki, using `newwikiname` with user `mediawiki` for the database?

And for uninstalling, I can just delete `/var/www/newwikiname` and `drop table newwikiname`?

**Reply by MichaIng** [reply]
Jep exactly that. For uninstall and reinstall, if there is no good safe file to identify MediaWiki, we could also create an own dotfile like `.dietpi-mediawiki` or something like that, so we do not accidentally remove other dirs.
